sympy/sets/fancysets.py

```diff
@@ -909,27 +936,34 @@ def _boundary(self):
 
     def as_relational(self, x):
         """Rewrite a Range in terms of equalities and logic operators. """
-        if self.size == 1:
-            return Eq(x, self[0])
-        elif self.size == 0:
-            return S.false
+        from sympy.core.mod import Mod
+        if self.start.is_infinite:
+            assert not self.stop.is_infinite  # by instantiation
+            a = self.reversed.start
         else:
-            from sympy.core.mod import Mod
-            cond = None
-            if self.start.is_infinite:
-                if self.stop.is_infinite:
-                    cond = S.true
-                else:
-                    a = self.reversed.start
-            elif self.start == self.stop:
-                cond = S.false  # null range
-            else:
-                a = self.start
-            step = abs(self.step)
-            cond = Eq(Mod(x, step), a % step) if cond is None else cond
-            return And(cond,
-                       x >= self.inf if self.inf in self else x > self.inf,
-                       x <= self.sup if self.sup in self else x < self.sup)
+            a = self.start
+        step = self.step
+        in_seq = Eq(Mod(x - a, step), 0)
+        ints = And(Eq(Mod(a, 1), 0), Eq(Mod(step, 1), 0))
+        n = (self.stop - self.start)/self.step
+        try:
+            if n == 0:
+                return S.EmptySet.as_relational(x)
+            if n == 1:
+                return And(Eq(x, a), ints)
+            a, b = self.inf, self.sup
+            range_cond = And(
+                x > a if a.is_infinite else x >= a,
+                x < b if b.is_infinite else x <= b)
+        except ValueError:  # inf/sup unknown
```

Comment:
It would be better not to catch `ValueError`. Lots of things can raise `ValueError`. Ideally it would not be explicitly raised in sympy code when a sympy-specific exception could be used instead.

Reply:
I wrapped only the sup/inf part in the try/except clause. This is the only error arising from these routines under the condition of working with a set for which n not in (0,1).